### DISCUSSION TITLE
is psycopg supposed to be more performant than psycopg2?

### DISCUSSION THREAD

**Discussion by zzzeek** [psycopg/psycopg#411] [Q&A]
We've been running some [profiling](https://gist.github.com/CaselIT/4d2259dbd9007b099d829726c71c6d0f) on psycopg / psycopg2 after noticing some degraded performance in some of our SQLAlchemy examples.       So far, on a per-execute basis it seems psycopg is not as fast as its predecessor psycopg2.     It doesn't impact us directly, since we are just the library, but we do field performance issues and if it's a known thing that psycopg is slower than psycopg2, we'd want to be able to say that. it's just I always had the impression that improved performance was one of the goals of psycopg.

here's results for a local run here:

```
$ python test3.py psycopg2 psycopg psycopg_async asyncpg
Running psycopg2
Insert done. Running 10000 queries
Run psycopg2 in 1.459726708009839 s
Running psycopg sync
Insert done. Running 10000 queries
Run psycopg in 2.2721266420558095 s
Running psycopg async
Insert done. Running 10000 queries
Run psycopg_async in 2.933384043048136 s
Running asyncpg
Insert done. Running 10000 queries
Run asyncpg in 1.930702666984871 s
```

results of script below on @CaselIT 's env which I think is windows:

```
(sa) [federico@Camera sqlalchemy]$ python pg_perf.py asyncpg
Running asyncpg
Insert done. Running 10000 queries
Run psycopg2 in 8.307029499999999 s
(sa) [federico@Camera sqlalchemy]$ python pg_perf.py asyncpg
Running asyncpg
Insert done. Running 10000 queries
Run asyncpg in 8.173984500000003 s
(sa) [federico@Camera sqlalchemy]$ python pg_perf.py psycopg_async
Running psycopg async
Insert done. Running 10000 queries
Run psycopg_async in 10.595968999999997 s
(sa) [federico@Camera sqlalchemy]$ python pg_perf.py psycopg
Running psycopg sync
Insert done. Running 10000 queries
Run psycopg in 8.901597100000004 s
(sa) [federico@Camera sqlalchemy]$ python pg_perf.py psycopg2
Running psycopg2
Insert done. Running 10000 queries
Run psycopg2 in 6.6013320000000135 s
```

here's the latest script:

```py
import asyncio
from contextlib import contextmanager
import sys
import random
import time


table = """
CREATE TABLE customer (
        id SERIAL NOT NULL,
        name VARCHAR(255),
        description VARCHAR(255),
        q INTEGER,
        p INTEGER,
        x INTEGER,
        y INTEGER,
        z INTEGER,
        PRIMARY KEY (id)
)
"""
drop = "DROP TABLE IF EXISTS customer"

ids = list(range(10_000))
tests = 10000

data = [
    dict(
        id=i,
        name="c%d" % i,
        description="c%d" % i,
        q=i * 10,
        p=i * 20,
        x=i * 30,
        y=i * 40,
    )
    for i in ids
]

insert = """
INSERT INTO customer (id, name, description, q, p, x, y) VALUES
(%(id)s, %(name)s, %(description)s, %(q)s, %(p)s, %(x)s, %(y)s)
"""

select = """
SELECT customer.id, customer.name, customer.description, customer.q,
    customer.p, customer.x, customer.y, customer.z
FROM customer
WHERE customer.id = %(id)s
"""

url = "postgresql://scott:tiger@localhost/test"


@contextmanager
def time_log(message):
    start = time.monotonic()
    yield
    end = time.monotonic()
    print(f"Run {message} in {end-start} s")


def run_psycopg2():
    print("Running psycopg2")
    import psycopg2

    with psycopg2.connect(url) as conn:
        with conn.cursor() as cursor:
            cursor.execute(drop)
            cursor.execute(table)
            cursor.executemany(insert, data)
        conn.commit()
    print(f"Insert done. Running {tests} queries")
    to_query = random.choices(ids, k=tests)
    with psycopg2.connect(url) as conn:
        with time_log("psycopg2"):
            for id_ in to_query:
                with conn.cursor() as cursor:
                    cursor.execute(select, {"id": id_})
                    cursor.fetchall()
                conn.rollback()

    with psycopg2.connect(url) as conn:
        with conn.cursor() as cursor:
            cursor.execute(drop)
        conn.commit()


def run_psycopg():
    print("Running psycopg sync")
    import psycopg

    with psycopg.connect(url) as conn:
        with conn.cursor() as cursor:
            cursor.execute(drop)
            cursor.execute(table)
            cursor.executemany(insert, data)
        conn.commit()
    print(f"Insert done. Running {tests} queries")
    to_query = random.choices(ids, k=tests)
    with psycopg.connect(url) as conn:
        with time_log("psycopg"):
            for id_ in to_query:
                cursor = conn.execute(select, {"id": id_})
                cursor.fetchall()
                cursor.close()
                conn.rollback()

    with psycopg.connect(url) as conn:
        with conn.cursor() as cursor:
            cursor.execute(drop)
        conn.commit()


async def run_psycopg_async():
    print("Running psycopg async")
    import psycopg

    conn: psycopg.AsyncConnection
    async with await psycopg.AsyncConnection.connect(url) as conn:
        async with conn.cursor() as cursor:
            await cursor.execute(drop)
            await cursor.execute(table)
            await cursor.executemany(insert, data)
        await conn.commit()
    print(f"Insert done. Running {tests} queries")
    to_query = random.choices(ids, k=tests)
    async with await psycopg.AsyncConnection.connect(url) as conn:
        with time_log("psycopg_async"):
            for id_ in to_query:
                cursor = await conn.execute(select, {"id": id_})
                await cursor.fetchall()
                await cursor.close()
                await conn.rollback()

    async with await psycopg.AsyncConnection.connect(url) as conn:
        async with conn.cursor() as cursor:
            await cursor.execute(drop)
        await conn.commit()


async def run_asyncpg():
    print("Running asyncpg")
    import asyncpg

    places = dict(
        id="$1", name="$2", description="$3", q="$4", p="$5", x="$6", y="$7"
    )
    a_insert = insert % places
    a_select = select % {"id": "$1"}

    conn: asyncpg.Connection = await asyncpg.connect(url)
    async with conn.transaction():
        await conn.execute(drop)
        await conn.execute(table)
        await conn.executemany(a_insert, [tuple(d.values()) for d in data])
    await conn.close()
    print(f"Insert done. Running {tests} queries")
    to_query = random.choices(ids, k=tests)
    conn: asyncpg.Connection = await asyncpg.connect(url)
    with time_log("asyncpg"):
        for id_ in to_query:
            tr = conn.transaction()
            await tr.start()
            await conn.fetch(a_select, id_)
            await tr.rollback()
    await conn.close()

    conn: asyncpg.Connection = await asyncpg.connect(url)
    async with conn.transaction():
        await conn.execute(drop)
    await conn.close()


for name in sys.argv[1:]:
    if name == "psycopg2":
        run_psycopg2()
    elif name == "psycopg":
        run_psycopg()
    elif name == "psycopg_async":
        if hasattr(asyncio, "WindowsSelectorEventLoopPolicy"):
            asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())
        asyncio.run(run_psycopg_async())
    elif name == "asyncpg":
        asyncio.run(run_asyncpg())
    else:
        print("unknown driver", sys.argv[1])
```

**Comment by dvarrazzo** [comment]
In most cases, psycopg 3 with a C speedup (either `psycopg[binary]` or `psycopg[c]`) is slightly more performing than `psycopg2`. Are you maybe running a pure python version? What does `print(psycopg.pq.__impl__)` say?

There are some specific cases in which there is a performance regression: arrays for instances (see #359).

I'll take a look at your profile example. Thank you very much!

**Reply by CaselIT** [reply]
On my case I was trying with binary and c with similar result. I think also on mike part. My time above were on wsl2 with postgres running in a container, so not the fastest setup, but running things locally on windows I get similar ratios

**Reply by zzzeek** [reply]
yes w/ the pure python version things are much slower (my runs above are with the psycopg_binary package).   

what's the rationale for the pure Python version?   it's very slow and still requires libpq is present in any case.   These days you can use [cibuildwheel](https://github.com/pypa/cibuildwheel) and every arch will be there, nobody needs to compile anything anymore.

**Reply by zzzeek** [reply]
Here's the pure python run for me:

```
Running psycopg2
Insert done. Running 10000 queries
Run psycopg2 in 1.3872911459766328 s
Running psycopg sync
Insert done. Running 10000 queries
Run psycopg in 3.322731139953248 s
Running psycopg async
Insert done. Running 10000 queries
Run psycopg_async in 4.32279699400533 s
Running asyncpg
Insert done. Running 10000 queries
Run asyncpg in 1.939840400009416 s
```

**Comment by CaselIT** [comment]
Just a bit of context: this is a pure sql version of a performance example in the sqlalchemy repository.
Mike was doing experiments and we were seeing very similar time ratios when using sqlalchemy so we tried to look if it was something in sqlalchemy that influenced the performance or if the results were similar when running things without sqlalchemy.

The equivalent example can be run using `python -m examples.performance short_selects  --dburl postgresql+psycopg://scott:tiger@localhost/test --test test_core_reuse_stmt` on a sqlalchemy checkout.

There are other performance examples in sqlalchemy but I'm not sure if they are relevant since many focus on the orm. @zzzeek are there other core tests that may be useful to test in a pure sql way?

**Comment by dvarrazzo** [comment]
So, I've found at least two things: one shortcoming caused by the way your test benchmark is written and some improvements that we can do.

The test program has the problem of having the rollback in the loop; because of that, psycopg cannot make use of prepared statements. The reason is that, if an object is dropped, or created and dropped because the creating transaction rolls back, a prepared plan might refer to the dropped object, causing internal errors in postgres.

Looking into profile traces, I have noticed that epoll() has a non-trivial overhead: see #414.

Some numbers I obtain:

With rollback:
- psycopg2: 1.85 sec
- psycopg 3 epoll: 3.12 sec
- psycopg 3 select: 2.97 sec

Without rollback:
- psycopg2: 1.09 sec
- psycopg 3 epoll: 1.16 sec
- psycopg 3 select: 1.10 sec

Note: these tests are with python 3.8; psycopg 3 uses generators internally, so it might benefit more than psycopg2 of using Python 3.10, because of the optimization around exception management happened in that version.

There might be more fine tuning available to do and more code which can be pushed to the C layer. I'll look into it in the coming days.

**Reply by CaselIT** [reply]
Thanks for the analysis. Glad that's at least somewhat useful.

Regarding the rollback in the loop, I think it's there in the sqlalchemy example since that was testing a single short select in a one off transaction. This is probably not representative of the classic usage though.

**Comment by pwtail** [comment]
@dvarrazzo  Something bad is happening with concurrency. Increasing the number of concurrent connections makes psycopg performance worse (!). I am talking about both psycopg and psycopg_async.

Here is the modified [script](https://gist.github.com/pwtail/5f056752fed2760bc78d3fea64b6f854) I used.
In the case of sync driver adding concurrency means multiple threads, in the async case - multiple coroutines.

Both psycopg2 and asyncpg behave as expected so using concurrency roughly doubles their performance

**Reply by dvarrazzo** [reply]
Thank you for the report. Open #448 to investigate.
